File: 剑指offer/60.n个骰子的点数.py
from typing import List


class Solution:
    def dicesProbability(self, n: int) -> List[float]:
        ans = []

        dp = [[0 for _ in range(6 * n + 1)] for _ in range(n + 1)]

        for i in range(1, 7):
            dp[1][i] = 1

        for i in range(2, n + 1):
            for j in range(i, i * 6 + 1):
                for k in range(1, 7):
                    if j - k > 0:
                        dp[i][j] += dp[i - 1][j - k]
                    else:
                        break

        for i in range(n, 6 * n + 1):
            ans.append(dp[n][i] * pow(1 / 6, n))

        return ans

File: 剑指offer/ 60.n个骰子的点数.cpp
#include <algorithm>
#include <vector>

using namespace std;
class Solution {
public:
    vector<double> dicesProbability(int n)
    {
        vector<vector<int>> dp(11, vector<int>(70));

        for (int i = 1; i <= 6; i++) {
            dp[1][i] = 1;
        }

        for (int i = 2; i <= n; i++) {
            for (int j = i; j <= 6 * i; j++) {
                for (int cur = 1; cur <= 6; cur++) {
                    if (j - cur <= 0) {
                        break;
                    }
                    dp[i][j] += dp[i - 1][j - cur];
                }
            }
        }

        vector<double> ans;

        for (int s = n; s <= 6 * n; s++) {
            ans.push_back(1.0 / pow(6, n) * dp[n][s]);
        }

        return ans;
    }
};